File: src/client/java/net/bittorn/towsclient/KeybindManager.java
package net.bittorn.towsclient;

import net.bittorn.towsclient.data.npc.NPCRegistry;
import net.bittorn.towsclient.data.npc.models.BasicNPCData;
import net.bittorn.towsclient.screens.DialogScreen;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.option.KeyBinding;
import net.minecraft.client.util.InputUtil;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import org.lwjgl.glfw.GLFW;

import java.io.File;

public class KeybindManager {
    public static void register() {
        KeyBinding interactBinding = KeyBindingHelper.registerKeyBinding(new KeyBinding("key.towsclient.interact", InputUtil.Type.KEYSYM, GLFW.GLFW_KEY_G, "key.category.towsclient"));

        ClientTickEvents.END_CLIENT_TICK.register(client -> {
            while (interactBinding.wasPressed()) {
                if (!TOWSClient.enabled) return;

                Identifier npcId = Identifier.of(TOWSClient.MOD_ID, "skyland" + File.separator + "placeholder_npc");

                try {
                    BasicNPCData npc = NPCRegistry.getOrEmpty(npcId).orElseThrow();

                    int interactionRange = 4;
                    assert client.player != null;

                    BlockPos playerBlockPos = client.player.getBlockPos();

//                  // placeholder code for regions
//                    boolean isWithinRangeX = (clientBlockPos.getX() >= -16 && clientBlockPos.getX() <= 16);
//                    boolean isWithinRangeZ = (clientBlockPos.getZ() >= -16 && clientBlockPos.getZ() <= 16);
//                    if (!isWithinRangeX || !isWithinRangeZ) {
//                        return;
//                    }

                    if (playerBlockPos.isWithinDistance(npc.getPosition(), interactionRange)) {
                        MinecraftClient.getInstance().setScreen(new DialogScreen(npc.getDialogId()));
                    }
                } catch (Exception e) {
                    TOWSClient.LOGGER.error("Error parsing NPC: {}", npcId, e);
                }
            }
        });
    }
}

File: src/main/kotlin/com/windtempos/tows/util/KeybindManager.kt
package com.windtempos.tows.util

import com.mojang.blaze3d.platform.InputConstants
import com.windtempos.tows.TalesOfWaywardStars
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper
import net.minecraft.client.KeyMapping
import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.resources.Identifier
import org.lwjgl.glfw.GLFW
import java.io.File

class KeybindManager {

    companion object {
        fun register() {

            TalesOfWaywardStars.LOGGER.info("Registering keybinds")

            val interactKey = KeyBindingHelper.registerKeyBinding(
                KeyMapping(
                    "key.tales-of-wayward-stars.interact",
                    InputConstants.Type.KEYSYM,  // The type of the keybinding, KEYSYM for keyboard, MOUSE for mouse.
                    GLFW.GLFW_KEY_G,  // The keycode of the key
                    KeyMapping.Category.MISC // The category of the key
                )
            )

            ClientTickEvents.END_CLIENT_TICK.register(ClientTickEvents.EndTick { client: Minecraft? ->
                while (interactKey.consumeClick()) {
                    if (!TalesOfWaywardStars.enabled) return@EndTick

                    val npcId: Identifier =
                        Identifier.fromNamespaceAndPath(
                            TalesOfWaywardStars.MOD_ID,
                            "skyland" + File.separator + "placeholder_npc"
                        )

                    try {
                        //val npc: BasicNPCData = NPCRegistry.getOrEmpty(npcId).orElseThrow()

                        val interactionRange = 4.0
                        checkNotNull(client?.player)

                        val playerBlockPos: BlockPos = client.player!!.blockPosition()

                        //                  // placeholder code for regions
//                    boolean isWithinRangeX = (clientBlockPos.getX() >= -16 && clientBlockPos.getX() <= 16);
//                    boolean isWithinRangeZ = (clientBlockPos.getZ() >= -16 && clientBlockPos.getZ() <= 16);
//                    if (!isWithinRangeX || !isWithinRangeZ) {
//                        return;
//                    }

                        //if (playerBlockPos.closerThan(npc.position, interactionRange)) {
                        //    client.setScreen(DialogScreen(npc.getDialogId()))
                        //}
                    } catch (e: Exception) {
                        TalesOfWaywardStars.LOGGER.error("Error parsing NPC: {}", npcId, e)
                    }
                }
            })

            TalesOfWaywardStars.LOGGER.info("Keybinds registered successfully")
        }
    }
}